// xr-app-loader/lib/reducer.js
'use strict';

Object.defineProperty(exports, "__esModule", {
    value: true
});

exports.default = function () {
    var state = arguments.length > 0 && arguments[0] !== undefined ? arguments[0] : (0, _immutable.Map)();
    var _ref = arguments[1];
    var type = _ref.type,
        payload = _ref.payload;

    switch (type) {
        case "@@loadAppReal":
            return loadApp(state, payload);
        case "@@reduce":
            return reduce(state, payload);
        case "@@clearAppState":
            return clearAppState(state, payload);
        default:
            return state;
    }
};

var _immutable = require('immutable');

var _wrapMapStateToProps = require('./wrapMapStateToProps');

var _wrapMapStateToProps2 = _interopRequireDefault(_wrapMapStateToProps);

var _wrapMapDispatchToProps = require('./wrapMapDispatchToProps');

var _wrapMapDispatchToProps2 = _interopRequireDefault(_wrapMapDispatchToProps);

var _createReduxConnector = require('./createReduxConnector');

var _createReduxConnector2 = _interopRequireDefault(_createReduxConnector);

function _interopRequireDefault(obj) { return obj && obj.__esModule ? obj : { default: obj }; }

function loadApp(state, _ref2) {
    var fullName = _ref2.fullName,
        _ref2$component = _ref2.component,
        component = _ref2$component === undefined ? {} : _ref2$component,
        _ref2$action = _ref2.action,
        action = _ref2$action === undefined ? {} : _ref2$action,
        _ref2$reducer = _ref2.reducer,
        reducer = _ref2$reducer === undefined ? {} : _ref2$reducer;

    var p = fullName.split('?'),
        name = p[0],
        query = p[1] || '';

    if (!state.has(name)) {
        state = state.set(name, (0, _immutable.Map)());
        state = state.setIn([name, '@@require'], (0, _immutable.Map)({
            component: component,
            action: action,
            reducer: reducer
        }));
    }
    if (!state.getIn([name, '@@require', fullName])) {
        var container = (0, _createReduxConnector2.default)(component, (0, _wrapMapStateToProps2.default)(fullName), (0, _wrapMapDispatchToProps2.default)(fullName, action, reducer), null, {
            withRef: true,
            pure: true
        });

        state = state.setIn([name, '@@require', fullName], container);
    }

    if (query !== '' && !state.get(name).has(query)) {
        state = state.update(name, function (x) {
            return x.set(query, (0, _immutable.Map)());
        });
    }

    return state;
}

function clearAppState(state, _ref3) {
    var fullName = _ref3.fullName;

    var p = fullName.split('?'),
        name = p[0],
        query = p[1] || '';

    if (!state.has(name)) return state;

    if (query !== '' && state.get(name).has(query)) {
        state = state.update(name, function (x) {
            return x.set(query, (0, _immutable.Map)());
        });
    } else {
        var ks = [];
        state.get(name).mapKeys(function (k) {
            if (k != '@@require' && k.indexOf('=') == -1) ks.push(k);
            return k;
        });

        ks.forEach(function (k) {
            if (k) state = state.update(name, function (x) {
                return x.remove(k);
            });
        });
    }
    return state;
}

function reduce(state, _ref4) {
    var reducer = _ref4.reducer,
        type = _ref4.type,
        payload = _ref4.payload,
        name = _ref4.name,
        query = _ref4.query,
        injectFunsForReducer = _ref4.injectFunsForReducer;

    var oldState = query !== '' ? state.getIn([name, query]) : state.get(name);
    var newState = reducer[type].apply(this, [oldState].concat(payload));
    if (typeof newState === "function") {
        newState = newState(injectFunsForReducer);
    }
    return query !== '' ? state.setIn([name, query], newState) : state.set(name, newState);
}
module.exports = exports['default'];
